File: leetcode/20.valid-parentheses/main.go
package main

func isValid(s string) bool {
	var stacks []byte

	for i := 0; i < len(s); i++ {
		if s[i] == '[' || s[i] == '{' || s[i] == '(' {
			stacks = append(stacks, s[i])
		} else if (s[i] == '}' && len(stacks) > 0 && stacks[len(stacks)-1] == '{') ||
			(s[i] == ']' && len(stacks) > 0 && stacks[len(stacks)-1] == '[') ||
			(s[i] == ')' && len(stacks) > 0 && stacks[len(stacks)-1] == '(') {
			stacks = stacks[:len(stacks)-1]
		} else {
			return false
		}
	}

	return len(stacks) == 0
}

func main() {

}

File: leetcode/20.valid-parentheses/main.rs
struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = vec![];
        for r in s.as_bytes() {
            match *r as _ {
                ']' => match stack.pop() {
                    Some(l) if l == '[' => {}
                    _ => return false,
                },
                '}' => match stack.pop() {
                    Some(l) if l == '{' => {}
                    _ => return false,
                },
                ')' => match stack.pop() {
                    Some(l) if l == '(' => {}
                    _ => return false,
                },
                c => stack.push(c),
            }
        }
        stack.len() == 0
    }
}

fn main() {
    println!("{:?}", Solution::is_valid("([]{})".into()));
}
